rustls/src/crypto/mod.rs

```diff
@@ -59,18 +59,30 @@ pub trait SupportedKxGroup: Send + Sync + Debug {
 pub trait ActiveKeyExchange: Send + Sync {
     /// Completes the key exchange, given the peer's public key.
     ///
-    /// The shared secret is passed into the `sink` function for further processing.
+    /// The shared secret is returned as a [`SharedSecret`] which can be constructed
+    /// from a `&[u8]`.
     ///
     /// This consumes and so terminates the [`ActiveKeyExchange`].
-    fn complete(
-        self: Box<Self>,
-        peer_pub_key: &[u8],
-        sink: &mut dyn FnMut(&[u8]),
-    ) -> Result<(), Error>;
+    fn complete(self: Box<Self>, peer_pub_key: &[u8]) -> Result<SharedSecret, Error>;
 
     /// Return the public key being used.
     fn pub_key(&self) -> &[u8];
 
     /// Return the group being used.
     fn group(&self) -> NamedGroup;
 }
+
+/// The result from `ActiveKeyExchange::complete` as a value.
+pub struct SharedSecret(Vec<u8>);
+
+impl From<&[u8]> for SharedSecret {
+    fn from(source: &[u8]) -> Self {
+        Self(source.to_vec())
+    }
+}
+
+impl AsRef<[u8]> for SharedSecret {
```

Comment:
How do you feel about doing some kind of zeroization here?

For the pki-types private key types I avoided `AsRef<[u8]>` in favor of an inherent `secret_der()` method, do you think it's worth doing something similar here? In particular, we could make the inherent method `pub(crate)` so that callers outside this crate can't access it, which might be nice?

Reply:
Yeah I still think #265 is worth doing for belt-and-braces, especially now we have more types that contain secrets (eg, `AeadKey`). This type is another candidate for that treatment.

>  which might be nice?

Yeah, agree -- have done that.

The other follow-on from this is to take the `SharedSecret` in the key schedule state transition that requires it; I think that clarifies the meaning of the code a bit.